packages/muelu/src/Transfers/GeneralGeometric/MueLu_GeneralGeometricPFactory_def.hpp

```diff
@@ -0,0 +1,843 @@
+//
+// ***********************************************************************
+//
+// @HEADER
+#ifndef MUELU_GENERALGEOMETRICPFACTORY_DEF_HPP
+#define MUELU_GENERALGEOMETRICPFACTORY_DEF_HPP
+
+#include <stdlib.h>
+#include <iomanip>
+
+
+#include <Teuchos_LAPACK.hpp>
+
+#include <Xpetra_CrsMatrixWrap.hpp>
+#include <Xpetra_ImportFactory.hpp>
+#include <Xpetra_Matrix.hpp>
+#include <Xpetra_MapFactory.hpp>
+#include <Xpetra_MultiVectorFactory.hpp>
+#include <Xpetra_VectorFactory.hpp>
+
+#include <Xpetra_IO.hpp>
+
+#include "MueLu_FactoryManagerBase.hpp"
+#include "MueLu_GeneralGeometricPFactory_decl.hpp"
+
+#include "MueLu_MasterList.hpp"
+#include "MueLu_Monitor.hpp"
+
+namespace MueLu {
+
+  template <class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
+  RCP<const ParameterList> GeneralGeometricPFactory<Scalar, LocalOrdinal, GlobalOrdinal, Node>::GetValidParameterList() const {
+    RCP<ParameterList> validParamList = rcp(new ParameterList());
+
+    //#define SET_VALID_ENTRY(name) validParamList->setEntry(name, MasterList::getEntry(name))
+    //SET_VALID_ENTRY("generalgeometric: coarsening rate");
+    //#undef  SET_VALID_ENTRY
+    // Coarsen can come in two forms, either a single char that will be interpreted as an integer which is used as the coarsening
+    // rate in every spatial dimentions, or it can be a longer string that will then be interpreted as an array of integers.
+    // By default coarsen is set as "2", hence a coarsening rate of 2 in every spatial dimension is the default setting!
+    validParamList->set< std::string>            ("Coarsen",                   "2", "Coarsening rate in all spatial dimensions");
+    validParamList->set< RCP<const FactoryBase> >("A",               Teuchos::null, "Generating factory of the matrix A");
+    validParamList->set< RCP<const FactoryBase> >("Nullspace",       Teuchos::null, "Generating factory of the nullspace");
+    validParamList->set< RCP<const FactoryBase> >("Coordinates",     Teuchos::null, "Generating factory for coorindates");
+
+    return validParamList;
+  }
+
+  template <class Scalar,class LocalOrdinal, class GlobalOrdinal, class Node>
+  void GeneralGeometricPFactory<Scalar, LocalOrdinal, GlobalOrdinal, Node>::DeclareInput(Level& fineLevel, Level& coarseLevel) const {
+    Input(fineLevel, "A");
+    Input(fineLevel, "Nullspace");
+
+    // check whether fine level coordinate information is available.
+    // If yes, request the fine level coordinates, coarse level coordinates
+    // will be generated by GeometricCoordinatesTransferFactory called by RAP factory
+    if (fineLevel.GetLevelID() == 0) {
+      if (fineLevel.IsAvailable("Coordinates", NoFactory::get())) {
+        fineLevel.DeclareInput("Coordinates", NoFactory::get(), this);
+      } else {
+	TEUCHOS_TEST_FOR_EXCEPTION(fineLevel.IsAvailable("Coordinates", NoFactory::get()),
+				   Exceptions::RuntimeError, "No Coordinates provided by the user.");
+      }
```

Comment:
Yes, we require the user to provide coordinates. 
But
`Input(fineLevel, "Coordinates");`
translates to
`fineLevel.DeclareInput("Coordinates", GetFactory("Coordinates").get(), this);`
which does not work on level 0.